### DISCUSSION TITLE
How to get IDE support

### DISCUSSION THREAD

**Discussion by hofbi** [aspect-build/rules_py#317] [Q&A] [answered]
Hi,

as part of your [README](https://github.com/aspect-build/rules_py?tab=readme-ov-file#differences), you mention

> Thanks to the virtualenv, you can open the project in an editor like PyCharm and have working auto-complete, jump-to-definition

It is not fully clear to me how to get there? Do I have to tell the IDE where to find the venv inside the sandbox?

**Comment by yannick-brunken-lt** [comment]
I'm also wondering how to do the IDE integration. Especially being able to use the debugger would be essential. Thanks!

**Comment by liningpan** [comment]
I've never really tried this, but I think the idea is to create a venv with everything using the `py_venv` rule. There was a bug before 0.7.x, where the created venv has invalid paths.

**Comment by hofbi** [comment]
Would the idea then be to have a `py_venv` target for the entire repo to provide the IDE venv?

**Comment by mattem** [comment, marked as answer]
A venv can be created for any `py_test` or `py_binary` target, via `bazel run //the/binary.venv`, rules_py will create a venv for that target in the directory next to the target.

> Would the idea then be to have a py_venv target for the entire repo to provide the IDE venv?

This really depends on a number of factors. If there is multiple Python dependency closures in the repo, or Python versions, then a single venv will cause conflicts. It also depends on the size of the repo, perhaps, as opening the root of a repo and attempting to install all of the dependencies would be too large, so a finer grained "per service" and "per target" approach is needed.
